### PR TITLE
Escape argument quotes and scripts passed to cmd

### LINKED ISSUE #16: Non cmd shells need their arguments escaped
`cmd` doesn't need to be quoted, but `sh`, `pwsh`, etc. do and they also need to be escaped😳

https://github.com/xt0rted/dotnet-run-script/blob/9f1110c87d242d53f592102f92fcaea703e6c69f/src/RunScriptCommand.cs#L117-L120

This should be fixed and published ahead of #10.

### PR DESCRIPTION
**Pull request by xt0rted**
This fixes three related bugs:

1. Scripts passed to `*sh` shells weren't being escaped which caused them to run unexpectedly or outright crash
2. Additional arguments passed after the `--` (double dash) had their quotes stripped which could result in unexpected results or outright crashes. This re-quotes them before passing them along with the rest of the script.
3. Scripts passed to `cmd.exe` risk incorrectly being interpreted due to quoting and escaping, so every character passed is now prefixed with `^` to [force literal interpretation](https://docs.microsoft.com/archive/blogs/twistylittlepassagesallalike/everyone-quotes-command-line-arguments-the-wrong-way).

A new `integration` job was also added to the `ci` workflow which takes the nupkg result from the `build` job and then runs through the `build` and `test` scripts on each os and shell combo. This is an attempt at trying to catch some of these escaping and os/shell specific bugs before they make it into a release. There should probably be a couple scripts that we run specifically to test these things such as:

- Test file prefix with spaces
- Test file output location with spaces

Fixes #16

### CODE REVIEW

File: src/ArgumentEscaper.cs [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+        var length = arg.Length;
+        var needsQuotes = length == 0 || ArgumentContainsWhitespace(arg);
+        var isQuoted = needsQuotes || IsSurroundedWithQuotes(arg);
+
+        if (needsQuotes)
+        {
+            sb.Append(Quote);
+        }
+
+        for (var i = 0; i < length; ++i)
```
Comment (aarondandy) [inline src/ArgumentEscaper.cs:115]: If you are more consistent with usages of length and args.Length I wonder if that may help with any bounds checks optimizations.

File: src/ArgumentEscaper.cs [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        {
+            var backslashCount = 0;
+
+            // Consume All Backslashes
+            while (i < arg.Length && arg[i] == Backslash)
```
Comment (aarondandy) [inline src/ArgumentEscaper.cs:120]: May be a bit faster, maybe 🤷‍♂️, while also improving legibility... maybe 🤣
```csharp
var backslashCount = countChars(arg, Backslash, i);
i += backslashCount;
static int countChars(string arg, char c, int i) {
  for (; i < arg.Length && arg[i] == c; i++) ;
  return i;
}
```

File: src/ArgumentEscaper.cs [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+    /// <remarks>
+    /// See here for more info: <seealso href="https://docs.microsoft.com/archive/blogs/twistylittlepassagesallalike/everyone-quotes-command-line-arguments-the-wrong-way"/>.
+    /// </remarks>
+    /// <param name="arguments">List of arguments to escape.</param>
+    /// <returns>An escaped string of the supplied arguments.</returns>
+    private static string EscapeArgArray(IReadOnlyList<string> arguments)
+    {
+        var sb = new ValueStringBuilder(stackalloc char[256]);
+
+        for (var i = 0; i < arguments.Count; i++)
```
Comment (aarondandy) [inline src/ArgumentEscaper.cs:66]: I'm curious if there is any performance difference with enumerator. Also, then you could just make this IEnumerable. You are already reaching in via an interface so probably won't do much harm.
```csharp
var enumerator = arguments.GetEnumerator();
if (enumerator.MoveNext()) {
  var sb = new ValueStringBuilder(stackalloc char[256]);
  EscapeSingleArg(ref sb, enumerator.Current);

  while (enumerator.MoveNext()) {
    sb.Append(Space);
    EscapeSingleArg(ref sb, enumerator.Current);
  }
}
return string.Empty;
```

**Review by aarondandy — APPROVED** [review]
:shipit: , but some ideas

File: src/ArgumentEscaper.cs [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+            sb.Append(Quote);
+        }
+    }
+
+    private static bool IsSurroundedWithQuotes(ReadOnlySpan<char> argument)
+        => argument[0] == Quote &&
+            argument[^1] == Quote;
+
+    private static bool ArgumentContainsWhitespace(ReadOnlySpan<char> argument)
+        => argument.Contains(Space) ||
```
[the remark this answers is not in the document]
Comment (aarondandy) [reply src/ArgumentEscaper.cs:204]: Oh shit, I forgot that there is also an IndexOfAny on ReadOnlySpan too:
```csharp
public static bool ContainsAny(this ReadOnlySpan<char> @this, char value0, char value1) => @this.IndexOfAny(value0, value1) >= 0;
```

File: src/ArgumentEscaper.cs [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+        if (quoted)
+        {
+            sb.Append(Caret);
+            sb.Append(Quote);
+        }
+    }
+
+    private static bool IsSurroundedWithQuotes(ReadOnlySpan<char> argument)
+        => argument[0] == Quote &&
```
Comment (aarondandy) [inline src/ArgumentEscaper.cs:200]: May be worth adding an `argument.Length >= 2` check

File: src/ValueStringBuilder.cs [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    public void Append(char c, int count)
+    {
+        if (_pos > _chars.Length - count)
+        {
+            Grow(count);
+        }
+
+        var dst = _chars.Slice(_pos, count);
+
+        for (var i = 0; i < dst.Length; i++)
```
Comment (aarondandy) [inline src/ValueStringBuilder.cs:79]:
```csharp
_chars.Slice(_pos, count).Fill(c);
```

File: src/ArgumentEscaper.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (xt0rted) [reply src/ArgumentEscaper.cs:204]: I'm targeting 3.1, 5.0, and 6.0 and in 3.1 this call looks to be handled by `System.MemoryExtensions.Contains()` unless VS is lying to me.

File: src/ArgumentEscaper.cs (hunk 1 above)
Comment (xt0rted) [reply src/ArgumentEscaper.cs]: That came from the original source, but I removed `length` in favor of `arg.Length` for consistency.

File: src/ArgumentEscaper.cs (hunk 5 above)
Comment (xt0rted) [reply src/ArgumentEscaper.cs:229]: I don't think it's possible to get a value here that's just a single quote. If there is I haven't been able to reproduce it via the command line. Single character arguments are fine but probably aren't worth adding another check when most are going to be multiple characters.

File: src/ArgumentEscaper.cs (hunk 3 above)
Comment (xt0rted) [reply src/ArgumentEscaper.cs]: Now that `args` is a `string[]` the `for` loop benchmarks the fastest here.

File: src/RunScriptCommand.cs [hunk 7]
```diff
@@ -114,10 +110,22 @@ public async Task<int> InvokeAsync(InvocationContext context)
...
 
-            process.StartInfo.Arguments =
-                isCmd
-                ? $"/d /s /c {toExecute}"
-                : $"-c \"{toExecute}\"";
+            if (isCmd)
+            {
+                process.StartInfo.Arguments = string.Concat(
+                    "/d /s /c \"",
+                    cmd,
```
Comment (xt0rted) [inline src/RunScriptCommand.cs:117]: This needs to be escaped with `^` just like the arguments

**Pull request merged**